README.md
<img src="/docs/stocksight.png?raw=true" alt="stocksight" />

[![License](https://img.shields.io/github/license/shirosaidev/stocksight.svg?label=License&maxAge=86400)](./LICENSE)
[![Release](https://img.shields.io/github/release/shirosaidev/stocksight.svg?label=Release&maxAge=60)](https://github.com/shirosaidev/stocksight/releases/latest)
[![Sponsor Patreon](https://img.shields.io/badge/Sponsor%20%24-Patreon-brightgreen.svg)](https://www.patreon.com/shirosaidev)
[![Donate PayPal](https://img.shields.io/badge/Donate%20%24-PayPal-brightgreen.svg)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=CLF223XAS4W72)

# stocksight
Stock market analyzer and stock predictor using Elasticsearch, Twitter, News headlines and Python natural language processing and sentiment analysis. How much do emotions on Twitter and news headlines affect a stock's price? Let's find out...

## About
stocksight is an open source stock market analysis software that uses Elasticsearch to store Twitter and news headlines data for stocks. stocksight analyzes the emotions of what the author writes and does sentiment analysis on the text to determine how the author "feels" about a stock.


## Slack workspace
Join the conversation, get support, etc on [stocksight Slack](https://join.slack.com/t/stocksightworkspace/shared_invite/enQtNzk1ODI0NjA3MTM4LTA3ZDA0YzllOGNiM2I5ZjAzYWM2MjNmMjI0OTRlY2ZjYTk1NmM5YmEwMmMwOTE2OTNiMGZlNzdjZmZkM2RjM2U).


## Requirements
- Python 3.x
- Elasticsearch 5.x
- Kibana 5.x
- elasticsearch python module
- nltk python module
- requests python module
- tweepy python module
- beautifulsoup4 python module
- textblob python module
- vaderSentiment python module
- newspaper3k python module

### Download

```shell
$ git clone https://github.com/shirosaidev/stocksight.git
$ cd stocksight
```
[Download latest version](https://github.com/shirosaidev/stocksight/releases/latest)

## Screenshot
Stocksight Kibana dashboard
<img src="https://github.com/shirosaidev/stocksight/blob/master/docs/stocksight-dashboard-kibana.png?raw=true" alt="stocksight kibana dashboard" />


## Install - Docker

*** **See [how to use](#how-to-use) below before building the Docker containers** ***

1) Download/clone stocksight repo with git.
2) Set up stocksight, elasticsearch and kibana containers using Docker compose
```
cd stocksight
cp config.py.sample config.py
***see how to use below for config.py (stocksight config) changes***
docker-compose build && docker-compose up
```
**This will volume mount config.py (stocksight settings) and twitteruserids.txt to those files in your local git cloned "stocksight" directory**

3) Once all the containers have started up, shell into the container

`docker exec -it stocksight_stocksight_1 bash`

4) See examples below for running stocksight.

## Install - local

**Recommended to install Elasticsearch and Kibana in local machine or other machine/vm/docker**

1) Install python requirements using pip

`pip install -r requirements.txt`

2) Install python nltk data

`python -c "import nltk; nltk.download('punkt'); nltk.download('stopwords')"`


## How to use
1) Create a new twitter application and generate your consumer key and access token. https://developer.twitter.com/en/docs/basics/developer-portal/guides/apps.html
https://developer.twitter.com/en/docs/basics/authentication/guides/access-tokens.html

2) Copy config.py.sample to config.py (stocksight config file)

3) Set elasticsearch settings in config.py for your env (for Docker, set `elasticsearch_host = "elasticsearch"`)

4) Add twitter consumer key/access token and secrets to config.py

5) Edit config.py and modify NLTK tokens required/ignored and twitter feeds you want to mine. NLTK tokens required are keywords which must be in tweet before adding it to Elasticsearch (whitelist). NLTK tokens ignored are keywords which if are found in tweet, it will not be added to Elasticsearch (blacklist).

### Examples

Run sentiment.py to create 'stocksight' index in Elasticsearch and start mining and analyzing Tweets using keywords and the stock symbol TSLA

```sh
$ python sentiment.py -s TSLA -k 'Elon Musk',Musk,Tesla,SpaceX --debug
```

Start mining and analyzing Tweets using keywords and the stock symbol TSLA and follow any url links in tweets and performing sentiment analysis on the link web page as well as the tweet

```sh
$ python sentiment.py -s TSLA -k 'Elon Musk',Musk,Tesla,SpaceX -l --debug
```

Start mining and analyzing Tweets from feeds in config using cached user ids from file (if you change any of the twitter  feeds in the config file, you need to delete this file and recreate it without -f)

```sh
$ python sentiment.py -s TSLA -f twitteruserids.txt --debug
```

Start mining and analyzing News headlines and following headline links and scraping relevant text on landing page

```sh
$ python sentiment.py -s TSLA --followlinks --debug
```

Run stockprice.py to add stock prices to 'stocksight' index in Elasticsearch

```sh
$ python stockprice.py -s TSLA --debug
```

### Kibana

Load 'stocksight' index in Kibana. For index pattern you can use 'stocksight' if you only have the single index or 'stocksight-*', etc. For time-field name you will want to use the date/time field 'date'.

To import the saved exported visualizations/dashboard, go to Kibana, click on management, click on saved objects, click on the import button and import the export.json file.


### CLI options

```
usage: sentiment.py [-h] [-i INDEX] [-d] -s SYMBOL [-k KEYWORDS] [-a] [-u URL]
                    [-f FILE] [-l] [-n] [--frequency FREQUENCY]
                    [--followlinks] [-w]
                    [--overridetokensreq TOKEN [TOKEN ...]]
                    [--overridetokensignore TOKEN [TOKEN ...]] [-v] [--debug]
                    [-q] [-V]

optional arguments:
  -h, --help            show this help message and exit
  -i INDEX, --index INDEX
                        Index name for Elasticsearch (default: stocksight)
  -d, --delindex        Delete existing Elasticsearch index first
  -s SYMBOL, --symbol SYMBOL
                        Stock symbol you are interesed in searching for,
                        example: TSLA
  -k KEYWORDS, --keywords KEYWORDS
                        Use keywords to search for in Tweets instead of feeds.
                        Separated by comma, case insensitive, spaces are ANDs
                        commas are ORs. Example: TSLA,'Elon
                        Musk',Musk,Tesla,SpaceX
  -a, --addtokens       Add nltk tokens required from config to keywords
  -u URL, --url URL     Use twitter users from any links in web page at url
  -f FILE, --file FILE  Use twitter user ids from file
  -l, --linksentiment   Follow any link url in tweets and analyze sentiment on
                        web page
  -n, --newsheadlines   Get news headlines instead of Twitter using stock
                        symbol from -s
  --frequency FREQUENCY
                        How often in seconds to retrieve news headlines
                        (default: 120 sec)
  --followlinks         Follow links on news headlines and scrape relevant
                        text from landing page
  -w, --websentiment    Get sentiment results from text processing website
  --overridetokensreq TOKEN [TOKEN ...]
                        Override nltk required tokens from config, separate
                        with space
  --overridetokensignore TOKEN [TOKEN ...]
                        Override nltk ignore tokens from config, separate with
                        space
  -v, --verbose         Increase output verbosity
  --debug               Debug message output
  -q, --quiet           Run quiet with no message output
  -V, --version         Prints version and exits
  
  
usage: stockprice.py [-h] [-i INDEX] [-d] [-s SYMBOL] [-f FREQUENCY] [-v]
                     [--debug] [-q] [-V]

optional arguments:
  -h, --help            show this help message and exit
  -i INDEX, --index INDEX
                        Index name for Elasticsearch (default: stocksight)
  -d, --delindex        Delete existing Elasticsearch index first
  -s SYMBOL, --symbol SYMBOL
                        Stock symbol to use, example: TSLA
  -f FREQUENCY, --frequency FREQUENCY
                        How often in seconds to retrieve stock data (default:
                        120 sec)
  -v, --verbose         Increase output verbosity
  --debug               Debug message output
  -q, --quiet           Run quiet with no message output
  -V, --version         Prints version and exits
  ```


CHANGELOG.md
# stocksight Change Log

## [0.1-b.12] = 2020-06-08
### changed
- removed --noelasticsearch cli arg option

## [0.1-b.11] = 2020-05-24
### changed
- removed stocksight web site uploading capability, site has been removed

## [0.1-b.10] = 2020-03-29
### added
- Dockerfile and docker-compose.yml for running in docker
### changed
- added random time delay between fetching tweets to reduce chance of getting Twitter 420 code (throttled/backoff)
### fixed
- using -k keywords no longer causes twitter user id's to be looked up
- fatal error when looking up and using twitter user ids

## [0.1-b.9] = 2019-10-27
### added
- -l --linksentiment cli arg - follow any tweet link urls and run sentiment analysis on those web pages
- requirement for newspaper3k python module to requirements.txt, install with pip

## [0.1-b.8] = 2019-10-25
### added
- -w --websentiment cli arg - Get sentiment results from text processing website
- improved nltk token processing - no longed needed to provide multiple case in nltk tokens in config
- requirement for nltk stopwords, install with python -c "import nltk; nltk.download('stopwords')"
### changed
- getting web sentiment results from text processing website is no longer default
- improved tweet text cleaning, sentiment algorithm and stocksight sentiment upload values

## [0.1-b.7] = 2019-10-24
### added
- check if running Python 3
- -U --upload - uploads sentiment to stocksight website (BETA) https://stocksight.diskoverspace.com/
- stocksight_token in config.py.sample, used for auth to upload to stocksight website, copy to your config
- nltk_min_tokens in config.py.sample, used to set minimum number of tokens required, copy to your config
- tweet/news headline count/filtered/ratio log output
- --noelasticsearch cli arg for not adding new docs to Elasticsearch
- -s stock symbol cli arg (required arg), this is the stock symbol name and also used as the "tag name" on the stocksight webiste when uploading sentiment data
- --overridetokensreq and --overridetokensignore cli args
- -a --addtokens cli arg to add nltk required tokens from config to keywords
### changed
- --newsheadlines no longer requires stock symbol, use -s to provide stock symbol
- nltk required tokens from config now do not automatically get added to keywords, use -a or --addtokens to add them
### fixed
- 'NoneType' object is not iterable Can't get sentiment from url caused by 400 Form Validation Errors text: This field is required traceback error when tweet with no text passed to sentiment_analysis

## [0.1-b.6] = 2019-07-15
### fixed
- "TypeError: sequence item 0: expected str instance, int found" traceback error when running with -f twitteruserids.txt

## [0.1-b.5] = 2019-01-11
### changed
- set encoding to utf-8 and checked for bytes when writing to twitteruserids.txt

## [0.1-b.4] = 2018-12-10
### fixed
- TypeError: can't concat str to bytes when writing to twitteruserids.txt

## [0.1-b.3] = 2018-11-23
### added
- requirements.txt for installing python requirements with pip
- config.py.sample has new setting for specifying elasticsearch host/ip, port, username and password, copy to your config file

## [0.1-b.2] = 2018-10-10
### added
- cli option -n --newsheadlines to fetch and analyze stock symbol headlines from yahoo finance website instead of twitter
- cli option --frequency to control how often news headlines are retrieved
- cli option --followlinks to follow any links in news headlines and scrape any relevant text on landing page
- additional mappings for newsheadline docs in elasticsearch indices
### changed
- code cleanup

## [0.1-b.1] = 2018-10-09
### note
- first release


stockprice.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""stockprice.py - get stock price from Yahoo and add to
Elasticsearch.
See README.md or https://github.com/shirosaidev/stocksight
for more information.

Copyright (C) Chris Park 2018-2020
stocksight is released under the Apache 2.0 license. See
LICENSE for the full license text.
"""

import time
import requests
import re
import argparse
import logging
import sys
from elasticsearch import Elasticsearch
from random import randint

# import elasticsearch host
from config import elasticsearch_host, elasticsearch_port, elasticsearch_user, elasticsearch_password

from sentiment import STOCKSIGHT_VERSION
__version__ = STOCKSIGHT_VERSION

# url to fetch stock price from, SYMBOL will be replaced with symbol from cli args
url = "https://query1.finance.yahoo.com/v8/finance/chart/SYMBOL?region=US&lang=en-US&includePrePost=false&interval=2m&range=5d&corsDomain=finance.yahoo.com&.tsrc=finance"

# create instance of elasticsearch
es = Elasticsearch(hosts=[{'host': elasticsearch_host, 'port': elasticsearch_port}],
                   http_auth=(elasticsearch_user, elasticsearch_password))

class GetStock:

    def get_price(self, url, symbol):
        import re

        while True:

            logger.info("Grabbing stock data for symbol %s..." % symbol)

            try:

                # add stock symbol to url
                url = re.sub("SYMBOL", symbol, url)
                # get stock data (json) from url
                try:
                    r = requests.get(url)
                    data = r.json()
                except (requests.HTTPError, requests.ConnectionError, requests.ConnectTimeout) as re:
                    logger.error("Exception: exception getting stock data from url caused by %s" % re)
                    raise
                logger.debug(data)
                # build dict to store stock info
                try:
                    D = {}
                    D['symbol'] = symbol
                    D['last'] = data['chart']['result'][0]['indicators']['quote'][0]['close'][-1]
                    if D['last'] is None:
                        D['last'] = data['chart']['result'][0]['indicators']['quote'][0]['close'][-2]
                    D['date'] = time.strftime('%Y-%m-%dT%H:%M:%S', time.gmtime())  # time now in gmt (utc)
                    try:
                        D['change'] = (data['chart']['result'][0]['indicators']['quote'][0]['close'][-1] -
                                       data['chart']['result'][0]['indicators']['quote'][0]['close'][-2]) / \
                                        data['chart']['result'][0]['indicators']['quote'][0]['close'][-2] * 100
                    except TypeError:
                        D['change'] = (data['chart']['result'][0]['indicators']['quote'][0]['close'][-2] -
                                       data['chart']['result'][0]['indicators']['quote'][0]['close'][-3]) / \
                                      data['chart']['result'][0]['indicators']['quote'][0]['close'][-3] * 100
                        pass
                    D['high'] = data['chart']['result'][0]['indicators']['quote'][0]['high'][-1]
                    if D['high'] is None:
                        D['high'] = data['chart']['result'][0]['indicators']['quote'][0]['high'][-2]
                    D['low'] = data['chart']['result'][0]['indicators']['quote'][0]['low'][-1]
                    if D['low'] is None:
                        D['low'] = data['chart']['result'][0]['indicators']['quote'][0]['low'][-2]
                    D['vol'] = data['chart']['result'][0]['indicators']['quote'][0]['volume'][-1]
                    if D['vol'] is None:
                        D['vol'] = data['chart']['result'][0]['indicators']['quote'][0]['volume'][-2]
                    logger.debug(D)
                except KeyError as e:
                    logger.error("Exception: exception getting stock data caused by %s" % e)
                    raise

                # check before adding to ES
                if D['last'] is not None and D['high'] is not None and D['low'] is not None:
                    logger.info("Adding stock data to Elasticsearch...")
                    # add stock price info to elasticsearch
                    es.index(index=args.index,
                             doc_type="stock",
                             body={"symbol": D['symbol'],
                                   "price_last": D['last'],
                                   "date": D['date'],
                                   "change": D['change'],
                                   "price_high": D['high'],
                                   "price_low": D['low'],
                                   "vol": D['vol']
                                   })
                else:
                    logger.warning("Some stock data had null values, not adding to Elasticsearch")

            except Exception as e:
                logger.error("Exception: can't get stock data, trying again later, reason is %s" % e)
                pass

            logger.info("Will get stock data again in %s sec..." % args.frequency)
            time.sleep(args.frequency)


if __name__ == '__main__':

    # parse cli args
    parser = argparse.ArgumentParser()
    parser.add_argument("-i", "--index", metavar="INDEX", default="stocksight",
                        help="Index name for Elasticsearch (default: stocksight)")
    parser.add_argument("-d", "--delindex", action="store_true",
                        help="Delete existing Elasticsearch index first")
    parser.add_argument("-s", "--symbol", metavar="SYMBOL",
                        help="Stock symbol to use, example: TSLA")
    parser.add_argument("-f", "--frequency", metavar="FREQUENCY", default=120, type=int,
                        help="How often in seconds to retrieve stock data (default: 120 sec)")
    parser.add_argument("-v", "--verbose", action="store_true",
                        help="Increase output verbosity")
    parser.add_argument("--debug", action="store_true",
                        help="Debug message output")
    parser.add_argument("-q", "--quiet", action="store_true",
                        help="Run quiet with no message output")
    parser.add_argument("-V", "--version", action="version",
                        version="stocksight v%s" % STOCKSIGHT_VERSION,
                        help="Prints version and exits")
    args = parser.parse_args()

    # set up logging
    logger = logging.getLogger('stocksight')
    logger.setLevel(logging.INFO)
    eslogger = logging.getLogger('elasticsearch')
    eslogger.setLevel(logging.WARNING)
    requestslogger = logging.getLogger('requests')
    requestslogger.setLevel(logging.WARNING)
    logging.addLevelName(
        logging.INFO, "\033[1;32m%s\033[1;0m"
                      % logging.getLevelName(logging.INFO))
    logging.addLevelName(
        logging.WARNING, "\033[1;31m%s\033[1;0m"
                         % logging.getLevelName(logging.WARNING))
    logging.addLevelName(
        logging.ERROR, "\033[1;41m%s\033[1;0m"
                       % logging.getLevelName(logging.ERROR))
    logging.addLevelName(
        logging.DEBUG, "\033[1;33m%s\033[1;0m"
                       % logging.getLevelName(logging.DEBUG))
    logformatter = '%(asctime)s [%(levelname)s][%(name)s] %(message)s'
    loglevel = logging.INFO
    logging.basicConfig(format=logformatter, level=loglevel)
    if args.verbose:
        logger.setLevel(logging.INFO)
        eslogger.setLevel(logging.INFO)
        requestslogger.setLevel(logging.INFO)
    if args.debug:
        logger.setLevel(logging.DEBUG)
        eslogger.setLevel(logging.DEBUG)
        requestslogger.setLevel(logging.DEBUG)
    if args.quiet:
        logger.disabled = True
        eslogger.disabled = True
        requestslogger.disabled = True

    # print banner
    if not args.quiet:
        c = randint(1, 4)
        if c == 1:
            color = '31m'
        elif c == 2:
            color = '32m'
        elif c == 3:
            color = '33m'
        elif c == 4:
            color = '35m'

        banner = """\033[%s
       _                     _                 
     _| |_ _           _   _| |_ _     _   _   
    |   __| |_ ___ ___| |_|   __|_|___| |_| |_ 
    |__   |  _| . |  _| '_|__   | | . |   |  _|
    |_   _|_| |___|___|_,_|_   _|_|_  |_|_|_|  
      |_|                   |_|   |___|                
          :) = +$   :( = -$    v%s
     https://github.com/shirosaidev/stocksight
            \033[0m""" % (color, STOCKSIGHT_VERSION)
        print(banner + '\n')

    # set up elasticsearch mappings and create index
    mappings = {
        "mappings": {
            "stock": {
                "properties": {
                    "symbol": {
                        "type": "keyword"
                    },
                    "price_last": {
                        "type": "float"
                    },
                    "date": {
                        "type": "date"
                    },
                    "change": {
                        "type": "float"
                    },
                    "price_high": {
                        "type": "float"
                    },
                    "price_low": {
                        "type": "float"
                    },
                    "vol": {
                        "type": "integer"
                    }
                }
            }
        }
    }

    if args.symbol is None:
        print("No stock symbol, see -h for help.")
        sys.exit(1)

    if args.delindex:
        logger.info('Deleting existing Elasticsearch index ' + args.index)
        es.indices.delete(index=args.index, ignore=[400, 404])

    logger.info('Creating new Elasticsearch index or using existing ' + args.index)
    es.indices.create(index=args.index, body=mappings, ignore=[400, 404])

    # create instance of GetStock
    stockprice = GetStock()

    try:
        # get stock price
        stockprice.get_price(symbol=args.symbol, url=url)
    except Exception as e:
        logger.warning("Exception: Failed to get stock data caused by: %s" % e)
    except KeyboardInterrupt:
        print("Ctrl-c keyboard interrupt, exiting...")
        sys.exit(0)
